scripts/prepare_lambda_deployment.py

```diff
@@ -0,0 +1,69 @@
+from argparse import ArgumentParser
+import json
+import logging
+from pathlib import Path
+import shutil
+import sys
+
+from azul import config
+from azul.files import write_file_atomically
+from azul.logging import configure_script_logging
+
+log = logging.getLogger(__name__)
+
+
+def transform_tf(input_json):
+    assert 'variable' not in input_json
+    input_json['variable'] = {'role_arn': {}}
+
+    input_json['output']['rest_api_id'] = {
+        'value': '${aws_api_gateway_rest_api.rest_api.id}'
+    }
+
+    for func in input_json['resource']['aws_lambda_function'].values():
+        assert func['source_code_hash'] == '${filebase64sha256("./deployment.zip")}'
+        func['source_code_hash'] = '${filebase64sha256("${path.module}/deployment.zip")}'
+        assert func['filename'] == "./deployment.zip"
+        func['filename'] = "${path.module}/deployment.zip"
+
+    # TODO: Remove when https://github.com/aws/chalice/issues/1237 is merged
+    # Terraform tries to remove the API Gateway deployment before removing the
+    # attached stage, which fails. Creating a new resource first allows
+    # Terraform to move the stage to the new resource before attempting to
+    # delete the old one.
+    assert 'lifecycle' not in input_json['resource']['aws_api_gateway_deployment']['rest_api']
+    input_json['resource']['aws_api_gateway_deployment']['rest_api']['lifecycle'] = {'create_before_destroy': True}
+
+    return input_json
+
+
+def main(argv):
+    parser = ArgumentParser(
+        description='Prepare the Terraform config generated by Chalice and '
+                    'copy it into the terraform/ directory.'
+    )
+    parser.add_argument('lambda_name', help='the lambda of the config that will be '
+                                            'transformed and copied')
+    options = parser.parse_args(argv)
+    source_dir = Path(config.project_root) / 'lambdas' / options.lambda_name / '.chalice' / 'terraform'
+    output_dir = Path(config.project_root) / 'terraform' / options.lambda_name
+    output_dir.mkdir(exist_ok=True)
+
+    deployment_src = source_dir / 'deployment.zip'
+    deployment_dst = output_dir / 'deployment.zip'
+    shutil.copyfile(deployment_src, deployment_dst)
+    log.info('Copied %s to %s', deployment_src, deployment_dst)
```

Comment:
Your default should be to log before an action, not after. That way the log gives context to any error messages and tracebacks that might be generated.

Reply:
Thanks for the explanation.